app/src/main/java/fi/aalto/legroup/achso/entities/OptimizedVideo.java

```diff
@@ -0,0 +1,359 @@
+    // Text is stored in (begin, end) pairs to annotationTextBuffer
+    // Author indices are to user pool
+    private long[] annotationTime;
+    private float[] annotationXY;
+    private int[] annotationTextStartEnd;
+    private int[] annotationAuthorUserIndex;
+
+    public UUID getId() {
+        return id;
+    }
+
+    public String getTitle() {
+        return title;
+    }
+
+    public void setTitle(String title) {
+        this.title = title;
+    }
+
+    public String getGenre() {
+        return genre;
+    }
+
+    public Uri getVideoUri() {
+        return Uri.parse(videoUri);
+    }
+
+    public Uri getThumbUri() {
+        return Uri.parse(thumbUri);
+    }
+
+    public Uri getManifestUri() {
+        return Uri.parse(manifestUri);
+    }
+
+    public Location getLocation() {
+        if (!hasLocation) {
+            return null;
+        }
+
+        Location location = new Location("optimized getter");
+        location.setLatitude(locationLatitude);
+        location.setLongitude(locationLongitude);
+        location.setAccuracy(locationAccuracy);
+        return location;
+    }
+
+    public String getTag() {
+        return tag;
+    }
+
+    public void setTag(String tag) {
+        this.tag = tag;
+    }
+
+    public int getAnnotationCount() {
+        return annotationTime.length;
+    }
+
+    public String getAnnotationText(int i) {
+
+        // Annotation text data is stored in one giant buffer and is referenced with indices
+        // The indices are stored in { Start0, End0, Start1, End1, ...} interleaved array
+        int start = annotationTextStartEnd[i * 2];
+        int end = annotationTextStartEnd[i * 2 + 1];
+
+        if (start < 0 || end < 0) {
+
+            // Mark null with negative indices
+            return null;
+        } else if (start == end) {
+
+            // If the range is empty it's better not to refer to the buffer unnecessarily so
+            // return an empty string
+            return "";
+        } else {
+
+            // This shouldn't create a copy of the buffer, and it doesn't at the moment. If it
+            // does somehow in some newer system need to tell Java to create a reference instead.
+            return annotationTextBuffer.substring(start, end);
+        }
+    }
+
+    public boolean isLocal() {
+        Uri videoUri = getVideoUri();
+
+        // Uris without a scheme are assumed to be local
+        if (videoUri.isRelative()) {
+            return true;
+        }
+
+        String scheme = videoUri.getScheme().trim().toLowerCase();
+
+        switch (scheme) {
+            case "file":
+            case "content":
+                return true;
+
+            default:
+                return false;
+        }
+    }
+
+    public boolean isRemote() {
+        return !isLocal();
+    }
+
+    /**
+     * Store the data of the video in a more GC friendly way.
+     */
+    public OptimizedVideo(Video video) {
```

Comment:
Also it's tightly coupled to the function below so the both would need to be lifted up in the file.